### PR TITLE
[dynamo] Support weakref objects

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128533

Fixes https://github.com/pytorch/pytorch/issues/125720

I was earlier worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then @zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls del x and x is the last reference).

So any code that relies on del x to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects.

The only corner case is when Dynamo is reconstructing the weakref object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.

Cpython notes
1) https://docs.python.org/3/library/weakref.html
2) https://docs.python.org/3/reference/datamodel.html#index-2

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -3247,13 +3292,13 @@ void install_tensor_aliasing_guard(
 
 void install_no_tensor_aliasing_guard(
     const py::list& guard_managers,
-    py::list tensor_names,
+    const py::list& tensor_names,
```
Comment (anijain2305) [inline torch/csrc/dynamo/guards.cpp:3295]: Linter error - unrelated to the PR

**Comment by williamwen42** [comment]
Add a test where the weakref'd object is deleted? e.g.
```python
x = Obj()
ref = weakref.ref(x)
opt_fn(ref, inp)
del x
opt_fn(ref, inp)
```

**Review by williamwen42 — COMMENTED** [review]
Add a test where the weakref'd object is deleted? e.g.
```python
x = Obj()
ref = weakref.ref(x)
opt_fn(ref, inp)
del x
opt_fn(ref, inp)
```

**Comment by anijain2305** [comment]
@williamwen42 added.

**Comment by williamwen42** [comment]
Is there an assumption that a weakref's call (i.e. `ref()`) won't change during tracing or in the middle of running a dynamo-optimized function?

**Comment by anijain2305** [comment]
@williamwen42 Nice catch! Yeah, this test fails

Let me think how to handle it

~~~
import torch
import weakref

def fn(y):
    x = torch.randn(4)
    x_weak = weakref.ref(x)
    if x_weak is not None and x_weak() is not None:
        z =  torch.sin(y)
    else:
        z = torch.cos(y)

    del x
    if x_weak is not None and x_weak() is not None:
        z =  torch.sin(y)
    else:
        z = torch.cos(y)
    return z




y = torch.ones(4)
ref = fn(y)

opt_fn = torch.compile(fn, backend="eager")
res = opt_fn(y)
print(ref, res)
~~~

**Comment by williamwen42** [comment]
I tried thinking about how to support weakrefs a few months ago and ran into the issue of determining when a weakref gets invalidated - it may be the case that weakrefs can be invalidated at anytime - perhaps by a different thread, or by the garbage collector? If we do manage to support detecting when weakrefs are invalidated, we also need to support tracing the finalizer/callback.

**Comment by anijain2305** [comment]
Lol ... this is tricky .. what was I thinking .. converting to draft

**Comment by mlazos** [comment]
From chatting offline I think it should be possible to support just getting the value out - that's most of the use cases I've seen.

Tracking the lifetime of the underlying object is much harder and I don't think it's needed.

**Review by jansel — APPROVED** [review]

**Comment by anijain2305** [comment]
@jansel @williamwen42 @mlazos 

I was worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then @zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls `del x` and x is the last reference). 

So any code that relies on `del x` to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects. 

The only corner case is when Dynamo is reconstructing the `weakref` object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**